eeps/eep-0066.md

```diff
@@ -0,0 +1,340 @@
+
+Having a distinct start character for this feature simplifies
+tokenizing and parsing.
+
+### Sigil
+
+In a general sense, a [Sigil][3], is a prefix to a variable
+that indicates its type, such as «`$I`» in Basic or Perl,
+where «`$`» is the sigil and «`I`» is the variable.
+
+Here we define a Sigil as a prefix (and a suffix) to a string literal
+that indicates how it should be interpreted.  The Sigil is
+a syntactic sugar that creates some Erlang term.
+
+A Sigil string literal consists of:
+
+1. The Sigil start character «`~`».
+2. The [Sigil Type][], a name that may be empty.
+3. The [String Content][] within [String Delimiters][].
+4. The [Sigil Suffix][], a name character sequence that may be empty.
+
+### Sigil Type
+
+Immediately following the [Sigil][] start character is the Sigil Type
+which is a name composed of a sequence of characters that are allowed
+as the second or later characters in a variable or an atom.
+In short ISO [Latin-1][] letters, digits, «`_`» and «`@`».
+The [Sigil Type][] may be empty.
+
+The Sigil Type defines how the [Sigil][] syntactic sugar
+shall be interpreted.  The suggested Sigil Types are:
+
+* «»: the Vanilla [Sigil][].
+
+  Creates a `unicode:unicode_binary()` in Erlang.
+  It is a string represented as a UTF-8 encoded binary,
+  equivalent to applying `unicode:characters_to_binary/1`
+  on the [String Content][].  The [String Delimiters][]
+  and escape characters work as for regular strings,
+  triple-quoted strings, or quoted atoms in Erlang.
+
+  So «`~"abc\d"`» is equivalent to «`<<"abc\d"/utf8>>`», and
+  «`~'abc"d'`» is equivalent to «`<<"abc\"d"/utf8>>`».
+
+* «`r`»: regular expression.
+
+  Creates a term `{RE::unicode:charlist(),Flags::[unicode:latin1_char()]}`
+  suitable as the `Regexp` argument to functions in the `re` module.
+  The `RE` element is the [String Content][],
+  and the `Flags` element is the [Sigil Suffix][].
+
+  The `re` module needs to be augmented with functions that
+  accept this tuple format containing both a regular expression,
+  and compile options in the guise of character flags.
+  Translation of the character `Flags` into `re:compile_option()`s
+  should be done by these `re` module functions.
+
+  Within the [String Content][], character escape sequences
+  are handled according to the regular expression rules,
+  and in addition to them the [end delimiter][4] can be escaped
+  with a «`\`» character.  Within triple-quote delimiters
+  according to [EEP 64][], end delimiter escaping is not done.
+
+### String Delimiters
+
+Immediately following the [Sigil Type][] is the string start delimiter.
+A specific start delimiter character has a corresponding
+end delimiter character.
+
+These are the start delimiter characters and they have themselves
+as end delimiter: single quote «`'`», and double quote «`"`».
+
+Triple-quote delimiters are also allowed, that is; a sequence of
+3 or more double quote «`"`» characters as described in [EEP 64][].
```

Comment:
I have re-read EEP 64 and it may be worth considering the meaning of verbatim strings with sigils. If `"""` are intended to be verbatim, then should `~"""` also be verbatim? If so, that would be the default mechanism of disabling escaping. If not, it is probably worth adding a line to this paragraph clarifying the relationship? :)

Reply:
My intention is, but since I reduced the scope of the EEP it got removed, that `~s"` should do escaping, and therefore so should `~s"""` since the `s` is explicit.  It would be confusing if `~s` behaved differently for `"` and for `"""`.

Then, also, `~S` should be verbatim, both for `"` and for `"""`.

I think an important use case would be making a UTF-8 binary out of a `"` or a `"""` string therefore it should be simple, as in just adding a `~` sigil to it.  So the proposal is that `~"` means `~s"` and `~"""` means `~S"""`.

The other possibility would be that `~` is a shorthand for `~s`, which would be implicit and contradict the other visible cue: `"` or `"""`, so it is hopefully intuitive that the quote decides.

Reply:
> I think an important use case would be making a UTF-8 binary out of a " or a """ string therefore it should be simple, as in just adding a ~ sigil to it. So the proposal is that ~" means ~s" and ~""" means ~S""".

Thanks. I think adding something along those lines would perfectly clarify it.